Add initial docs for sync in push (#773)
WalletConnect/walletconnect-docs: docs/specs/clients/push/usage-of-sync-api.md

# Usage of Sync API

In order to provide state synchronization between user's devices we must add [Sync API](../core/sync/readme.md) support. As per Sync API requierments user [must sign additional message](../core/sync/sync-protocol.md#generating-a-message-to-sign) to enable state synchronization. 

## Stores 

State synchronization is achieved by sharing the same key value stores among every client that user
registered in Push API. For Push API there are following stores:

> ### Push Subscription Store

Reviewer: Docs are missing store name. I suggest sth like `com.walletconnect.notify.pushSubscription`

#### Store Key

Key of [Push Subscription Store](#push-subscription-store) must be equal to `topic` from
[`PushSubscription`](#pushsubscription) data structure. 

#### PushSubscription

```typescript
  interface PushSubscription {
    topic: string;
    account: string;
    relay: RelayerTypes.ProtocolOptions;
    metadata: Metadata;
    scope: ScopeMap;
    expiry: number;
    selfPublicKey: string;
    dappPublicKey: string;
  }
```

> ### Push Proposal Store

Reviewer: Docs are missing store name. I suggest sth like `com.walletconnect.notify.pushProposal`

#### Pre-requisities
The Push subscribe flow will require a dapp to host a static json file which will contain a DID document compliant with `did:web` method as specified [here](https://w3c-ccg.github.io/did-method-web/). In this DID document we will specify a X25519 public key that will be used by the Push API protocol to derive a symmetric key for the Push topic.

#### Protocol
Subscribe protocol will be established as follows:

1. W1 fetches public key X from the did:web document
2. Subscribe topic is derived from the sha256 hash of public key X
3. W1 generates key pair Y
4. W1 derives symmetric key S with keys X and Y
5. W1 sends push subscribe request (type 1 envelope) on subscribe topic with subscriptionAuth
7. Response topic is derived from the sha256 hash of symmetric key S
8. W1 subscribes to response topic
10. Cast Server receives push subscribe request on subscribe topic
11. Cast Server derives symmetric key and decrypts subscriptionAuth
12. Cast Server triggers webhook to notify Dapp of new registered address
13. Cast Server generates key pair Z
14. Cast Server derives symmetric key P with keys Y and Z

> 15. Cast Server responds to push subscribe request on response topic

Author: Cast is the name that is currently used, copied over from the original protocol spec

> 16. W1, Wx receives push subscribe response on the response topic

Reviewer 2: Since `Wx` only syncs the subscription after it's settled it doesn't actually end up dealing with the response topic does it? Seems like the odd one out since everything else until `20.` (i.e the sync step) is `W1` only

17. Wallet derives symmetric key P
18. Push topic is derived from the sha256 hash of the symmetric key P

> 19. W1 syncs the push subscription with Wx

Reviewer: This should be done after step 5:
> W1 sends push subscribe request (type 1 envelope) on subscribe topic with subscriptionAuth

Author: I don't think so. As I understand the protocol, the subscription is actually stored and ready for PNs after the push subscribe response is received

Reviewer: If we don't sync response topic before step 16. the `Wx` clients cannot receive push subscribe response and will depend on W1 to always be there to receive the response

> 20. W1 subscribes to push topic for future push messages

Author: Makes sense, it would avoid false positives if subscription fails for some reason

# Usage of Sync API

> In order to provide state synchronization between user's devices we must add [Sync API](../core/sync/readme.md) support. As per Sync API requierments user [must sign additional message](../core/sync/sync-protocol.md#generating-a-message-to-sign) to enable state synchronization. 

Reviewer 2: ```suggestion
In order to provide state synchronization between a user's devices we must add [Sync API](../core/sync/readme.md) support. As per Sync API requirements, the user [must sign an additional message](../core/sync/sync-protocol.md#generating-a-message-to-sign) to enable state synchronization. 
```

## Stores 

> State synchronization is achieved by sharing the same key value stores among every client that user

Reviewer 2: ```suggestion
State synchronization is achieved by sharing the same key value stores among every client that the user
```

> registered in Push API. For Push API there are following stores:

Reviewer 2: ```suggestion
registered in Push API. For Push API there are the following stores:
```

### Push Subscription Store

#### Store Name
`com.walletconnect.notify.pushSubscription`

#### Store Key

Key of [Push Subscription Store](#push-subscription-store) must be equal to `topic` from
[`PushSubscription`](#pushsubscription) data structure. 

#### PushSubscription

```typescript
interface PushSubscription {
    topic: string;
    account: string;
    relay: RelayerTypes.ProtocolOptions;
    metadata: Metadata;
    scope: ScopeMap;
    expiry: number;
    selfPublicKey: string;
    selfPrivateKey: string;
    dappPublicKey: string;
}
```

### Push Proposal Store

#### Store Name
`com.walletconnect.notify.pushProposal`

#### Store Key

Key of [Push Proposal Store](#push-proposal-store) must be equal to `responseTopic` from
[`PushProposal`](#pushproposal) data structure. 

#### PushProposal

```typescript
interface PushProposal {
    publicKey: string;
    privateKey: string;
    responseTopic: stirng;
    metadata: PushClientTypes.Metadata;
    account: string;
    scope: string[];
 }
 ```

### Push Subscription Protocol in multiclient environment

#### Definitions

> - W1 - clients that have access W (Wallet) blockchain account keys.

Reviewer 2: ```suggestion
- W1 - clients that have access to W (Wallet) blockchain account keys.
```

> - Wx - peer clients that have access W (Wallet) blockchain account keys. 

Reviewer 2: ```suggestion
- Wx - peer clients that have access to W (Wallet) blockchain account keys. 
```

#### Pre-requisities
The Push subscribe flow will require a dapp to host a static json file which will contain a DID document compliant with `did:web` method as specified [here](https://w3c-ccg.github.io/did-method-web/). In this DID document we will specify a X25519 public key that will be used by the Push API protocol to derive a symmetric key for the Push topic.

#### Protocol
Subscribe protocol will be established as follows:

1. W1 fetches public key X from the did:web document
2. Subscribe topic is derived from the sha256 hash of public key X
3. W1 generates key pair Y
4. W1 derives symmetric key S with keys X and Y
5. W1 sends push subscribe request (type 1 envelope) on subscribe topic with subscriptionAuth
7. Response topic is derived from the sha256 hash of symmetric key S

> 8. W1 subscribes to response topic

Reviewer 2: Minor: steps jump from `8` to `10` here

9. Cast Server receives push subscribe request on subscribe topic
10. Cast Server derives symmetric key S and decrypts subscriptionAuth

> 11. Cast Server triggers webhook to notify Dapp of new registered address

Reviewer 3: might not be needed, but should we specify that Cast derives the response topic similar to #7 since the response topic is referenced on #14?

Author: I'm sticking to how it was defined in the original doc, it'd be confusing if we only changed it here

13. Cast Server generates key pair Z
14. Cast Server derives symmetric key P with keys Y and Z
15. Cast Server responds to push subscribe request on response topic
16. W1, Wx receives push subscribe response on the response topic

> 17. Wallet derives symmetric key P

Reviewer 2: ```suggestion
17. W1 derives symmetric key P
```

17. Push topic is derived from the sha256 hash of the symmetric key P
18. W1 subscribes to push topic for future push messages
19. W1 syncs the push subscription with Wx
20. Wx stores the push subscription
21. Wx subscribe to push topic coming from the sync request.

### Push Proposal Protocol in multiclient environment

> **WIP**: It is unclear whether or not we want to sync push proposals

Reviewer 3: For MVP, I think we should leave out push proposals since there's still a chance we'll be remove it

# Usage of Sync API

In order to provide state synchronization between a user's devices we must add [Sync API](../core/sync/readme.md) support. As per Sync API requirements, the user [must sign an additional message](../core/sync/sync-protocol.md#generating-a-message-to-sign) to enable state synchronization. 

## Stores 

State synchronization is achieved by sharing the same key value stores among every client that the user
registered in Push API. For Push API there are the following stores:

### Push Subscription Store

#### Store Name
`com.walletconnect.notify.pushSubscription`

#### Store Key

Key of [Push Subscription Store](#push-subscription-store) must be equal to `topic` from
[`PushSubscription`](#pushsubscription) data structure. 

#### PushSubscription

```typescript
interface PushSubscription {
    topic: string;
    account: string;
    relay: RelayerTypes.ProtocolOptions;
    metadata: Metadata;
    scope: ScopeMap;
    expiry: number;
    selfPublicKey: string;

>     selfPrivateKey: string;

Reviewer 4: why we couldn't sync `symKey` only as we're doing for chat thread?

Reviewer 4: I'm talking about symmetric key P to be able to decrypt push massages

Author: I'm considering it - probably will end up doing that but wanted to get input from the team

Author: Let's sync symkey only

    metadata: Metadata;
    scope: ScopeMap;
    expiry: number;

>     symkey: string;

Reviewer 2: Aligning with how we generally case this across specs/implementations:

```suggestion
    symKey: string;
```
